common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -428,19 +443,63 @@ public StyledText substring(int beginIndex, int endIndex) {
                 int startIndexInPart = Math.max(0, beginIndex - currentIndex);
                 int endIndexInPart = Math.min(part.length(), endIndex - currentIndex);
 
-                String includedSubstring =
-                        part.getString(null, PartStyle.StyleType.NONE).substring(startIndexInPart, endIndexInPart);
+                String fullString = part.getString(previousPartStyle, styleType);
 
-                includedParts.add(new StyledTextPart(
+                String beforeSubstring = fullString.substring(0, startIndexInPart);
+                String includedSubstring = fullString.substring(startIndexInPart, endIndexInPart);
+
+                // If the substring splits a formatting code, then we need to throw an exception
+                if (beforeSubstring.endsWith(String.valueOf(ChatFormatting.PREFIX_CODE))
+                        || includedSubstring.endsWith(String.valueOf(ChatFormatting.PREFIX_CODE))) {
+                    throw new IllegalArgumentException("The substring splits a formatting code.");
+                }
+
+                // Reparse the string to drop the formatting codes and keep them as part styles
+                includedParts.addAll(StyledTextPart.fromCodedString(
                         includedSubstring, part.getPartStyle().getStyle(), null, Style.EMPTY));
             }
 
-            currentIndex += part.length();
+            currentIndex += part.getString(previousPartStyle, styleType).length();
+
+            previousPartStyle = part.getPartStyle();
         }
 
         return new StyledText(includedParts, clickEvents, hoverEvents);
     }
 
+    public StyledText[] separate(int... indexes) {
+        return separate(PartStyle.StyleType.NONE, indexes);
+    }
+
+    /**
+     * Splits this {@link StyledText} into multiple {@link StyledText}s at the given indexes.
+     * @param styleType the style type to use when calculating indexes
+     * @param indexes the indexes to split at, in ascending order
+     * @return the split {@link StyledText}s as an array
+     */
+    public StyledText[] separate(PartStyle.StyleType styleType, int... indexes) {
```

Comment:
I were imagining a "split sibling" that would just keep the separating part as well. This is probably more clear and useful; thanks! 

However, I wonder if not "partition" would be a better name for this functionality.

Reply:
Hm, you decide. Either name is fine by me.

Reply:
Then I decide "partition", since this is what this function does :)